### PR TITLE
Merge rustc changes back into the miri repo

### PR DESCRIPTION
**Pull request by oli-obk**
This uses https://github.com/josh-project/josh (RIIR ftw)

The first time it looks a bit odd with the merges and so, all future syncs will work just like the instructions say. I'll add instructions for the sync from miri to rustc, too, starting to do that right now.

Initial cache building now takes 5 mins 🥳

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -275,3 +275,28 @@ see <https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html>.
...
+docker run -p 8000:8000 -e JOSH_REMOTE=https://github.com -v josh-vol:/data/git joshproject/josh-proxy:latest
+```
+
+Register the josh proxy as a remote
+
+```
+git remote add josh http://localhost:8000/rust-lang/rust.git:/src/tools/miri.git
+```
+
+Fetch (takes ca 5-10 min time the first time)
```
Comment (RalfJung) [inline CONTRIBUTING.md:294]:
```suggestion
Fetch (takes ca 5-10 min the first time)
```

**Comment by RalfJung** [comment]
I don't quite recognize our use-case in what Josh advertises itself for but if it works :shrug: 

I'm not super happy about having to run a docker container for the sync, let's hope the infra team automates this soon :)

**Comment by RalfJung** [comment]
> I'll add instructions for the sync from miri to rustc, too, starting to do that right now.

FWIW I did that earlier with git-subtree while testing https://github.com/rust-lang/miri/pull/2564 and it worked very well.

**Comment by oli-obk** [comment]
> > I'll add instructions for the sync from miri to rustc, too, starting to do that right now.
> 
> FWIW I did that earlier with git-subtree while testing #2564 and it worked very well.

yea, but I wanna make sure that josh and git-subtree work fine, and we don't make rustc-to-miri syncs harder due to git subtree in the other direction being weird (I don't think it is, but need to try it out)

**Comment by RalfJung** [comment]
Looks like bors is on strike again so I guess that gives us time to at least fix the typo I noted. ;)
@bors r-

**Comment by oli-obk** [comment]
> I don't quite recognize our use-case in what Josh advertises itself for but if it works 🤷
> 
> I'm not super happy about having to run a docker container for the sync, let's hope the infra team automates this soon :)

that's just me being lazy and following the instructions. I belive a `cargo install` plus a short shell script may suffice, too

**Comment by RalfJung** [comment]
@oli-obk bors seems to work again... why did you r- this?

**Comment by oli-obk** [comment]
The other PR was merged and this one didn't start

**Pull request merged**